### PR TITLE
DVDDemuxFFmpeg: Fix issues after st->cur_pts replacement

### LINKED ISSUE #23199: PVR recordings issues with Omega
### Describe the bug

I have PVR issues on Omega/ffmpeg 6.
Resuming recordings doesn't work (the elapsed time is garbage and audio/video skips).

I've diffed detailed logging and first difference that seems significant is:
```
ffmpeg[0x353d868X]: [mpegts] read_seek: [0 5439510000](tel:05439510000)
ffmpeg[0x353d868X]: [mpegts] gen_seek: [0 5439510000](tel:05439510000)
```
on ffmpeg 4 and:
```
ffmpeg[0x351cea50]: [mpegts] read_seek: 0 -9223372036800810000
ffmpeg[0x351cea50]: [mpegts] gen_seek: 0 -9223372036800810000
```
on ffmpeg 6 (presumably a no pts value).

That invalid timestamp seems to propagate through VideoPlayer. e.g.
```CDVDMsg::GENERAL_RESYNC(9223372036854365184.000000)```
and
```PrepareNextRender: frameOnScreen: 600001839.741000 renderPts: 600068506.412753 nextFramePts:9223372036854775808.000000 -> diff: -9223372036254707712.000000  render: false forceNext: true")```

### To Reproduce
I can reproduce with a ts file played as a recording through PVR Demo app.
[example file](https://drive.google.com/uc?id=1fKOWcRZhYbBxJifC7XO7fAJ4L10QPN7R&export=download)

If I play, seek to 10 minutes then stop, then try to play and resume, ffmpeg 4.3.2 is good. ffmpeg 6.0 is bad
It shows elapsed time as 577014:32:43/52.03 and it constantly skips with "large audio sync error").

It doesn't matter what hardware acceleration is used.
The failure occurs with software decode and egl render, and also with DRMPrime.

### Debuglog

[Working log file with ffmpeg 4](http://sprunge.us/ucezmf)
[Broken log file with ffmpeg 6](http://sprunge.us/5PWP0e)

### Screenshots 
Here are some links or screenshots to help explain the problem:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 21

The logs are from my own builds of LibreELEC on Pi4.
I also see the issue in LE nightly builds.

I don't have an exact timeframe when the issue started, but it's been present for a number of weeks.
Most likely coinciding with one of the ffmpeg version bumps.

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by popcornmix**
## Description
Fix issues after st->cur_pts replacement

## Motivation and context
The move to ffmpeg5 removed st->cur_pts which kodi was using in a number of places.

The fixup replaced these uses with m_pkt.pkt.dts,
but that makes no sense in the Seek code as
m_pkt.pkt is only updated in Read function,
so we are using a stale value.

Also m_pkt.pkt.dts may be AV_NOPTS_VALUE in
TransportStreamAudioState/TransportStreamVideoState

which needs protecting against, otherwise we perform arithmetic on that and that breaks playback.

## How has this been tested?
I've been running with it for a few days, and resuming from PVR recordings and seeking is fixed.
Live TV and playback from local files seem fine.

## What is the effect on users?
Fixes broken PVR recording resume
See: https://github.com/xbmc/xbmc/issues/23199

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by popcornmix** [comment]
I can't claim to know this code well enough to say this fix is correct and will cause no regressions.
But I can say it is much better than current kodi master if you use PVR and recordings.

So, I'll put this out as a starting point for discussion.

@FernetMenta I'm happy to hear any suggestions.

**Comment by bkuhls** [comment]
At first sight your patch looks good and fixes the problem I observed here when restarting .ts files at their previous position, which is broken in master branch.

**Comment by FernetMenta** [comment]
> I can't claim to know this code well enough to say this fix is correct and will cause no regressions. But I can say it is much better than current kodi master if you use PVR and recordings.
> 
> So, I'll put this out as a starting point for discussion.
> 
> @FernetMenta I'm happy to hear any suggestions.

I don't think this is correct. I am sure it breaks seeking through chapters because  m_currentPts is not set anymore after seek. Changing m_pkt.pkt.dts to st->start_time is not correct either. Thise are two different things. That will break many pvr streams.
Why did't you follow the strategy I proposed on the issue? I think this was reading out a packet after a seek.

**Comment by arthur-liberman** [comment]
Hi, I just tested this patch in IPTV catchup.
Kodi crashes and restarts immediately after seek back or forward.

**Comment by popcornmix** [comment]
> Hi, I just tested this patch in IPTV catchup.
Kodi crashes and restarts immediately after seek back or forward.

are you saying without this PR you don't have the crash?
Can you post the crash log?

**Comment by arthur-liberman** [comment]
> > Hi, I just tested this patch in IPTV catchup.
> > Kodi crashes and restarts immediately after seek back or forward.
> 
> are you saying without this PR you don't have the crash? Can you post the crash log?

I may have jumped the gun a bit, as I tested it on CoreELEC.
Is there a build for Windows that includes this fix for me to test it?

**Comment by popcornmix** [comment]
>Why did't you follow the strategy I proposed on the issue? I think this was reading out a packet after a seek.

Are you suggesting after seeking doing something similar to [this](https://github.com/xbmc/xbmc/blob/c1e6da41be1d41ea8d31e2d6bff024c23d127875/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L1288-L1308) in the non-`m_checkTransportStream` case?
i.e. calling `Read()` until `m_pkt.pkt.dts != AV_NOPTS_VALUE` (possibly either dts or pts?)

Is it a concern that the initial `Read()` packets with `AV_NOPTS_VALUE` are discarded, or do we assume that any I frame will have a valid pts/dts and we don't care about discarding non-i frames after a seek?

**Comment by FernetMenta** [comment]
Sorry, seems I haven't looked close enough. Could you make your debug logs from the issue available. I can't access them.

**Comment by FernetMenta** [comment]
> Are you suggesting after seeking doing something similar to [this](https://github.com/xbmc/xbmc/blob/c1e6da41be1d41ea8d31e2d6bff024c23d127875/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L1288-L1308) in the non-`m_checkTransportStream` case?

Yes, but slightly different. unref the packet and set result to -1 before the read. That way the packet is kept for the next read by player.

**Comment by popcornmix** [comment]
@FernetMenta I've updated to something closer to your view (if I've understood correctly).
Initial playback and seeking of a PVR recording works.
Stop and resume does play okay,  but the resume offset is lost.
e.g. play to 10m, stop and then resume, and the progress bar starts at 0m (although actual video played is from 10m).

I think that is because `m_startTime` is getting set in `CDVDDemuxFFmpeg::WaitForStreamReady` during the initial seek, and then `ConvertTimestamp` subtracts `m_startTime` from all timestamps meaning they are all zero based.

Might be useful if you give some direction in case I'm going down the wrong path.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 1]
```diff
@@ -1287,25 +1311,8 @@ bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double* startpts)
 
   if (m_checkTransportStream)
   {
-    XbmcThreads::EndTime<> timer(1000ms);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1290]: This block is really only intended for ts streams.

**Comment by FernetMenta** [comment]
You have to make sure that after a seek any saved packet is released.

> EDIT: Maybe consider this also for the m_checkTransportStream case

This case was before a seek, so it was ok to drop the packet. 
I think sometimes it is better to just duplicate a few lines of code than trying to reuse code. I would first focus on the transport stream case and test if the changes to TransportStreamAudio/VideoState() fixes the original issue.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 2]
```diff
@@ -2342,10 +2354,14 @@ TRANSPORT_STREAM_STATE CDVDDemuxFFmpeg::TransportStreamAudioState()
     for (unsigned int i = 0; i < m_pFormatContext->programs[m_program]->nb_stream_indexes; i++)
     {
       int idx = m_pFormatContext->programs[m_program]->stream_index[i];
+      // have we already seeked to this stream
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:2357]: did we? this block should be execture also for normal start of playback where no seek is involved. Reusing m_seekStream seems correct but the comment is misleading

**Comment by popcornmix** [comment]
@FernetMenta Try 3.

This seems to fix the original issue of resuming from correct point of a PVR recording.
Seeking seems okay. Seeking chapters in an ISO file also seemed to work.

**Comment by bkuhls** [comment]
> This seems to fix the original issue of resuming from correct point of a PVR recording. Seeking seems okay. Seeking chapters in an ISO file also seemed to work.

New commits still look good and still fix the bug with .ts files I saw: https://github.com/xbmc/xbmc/pull/23557#issuecomment-1654563775

**Comment by FernetMenta** [comment]
> @FernetMenta Try 3.
> 
> This seems to fix the original issue of resuming from correct point of a PVR recording. Seeking seems okay. Seeking chapters in an ISO file also seemed to work.

@popcornmix I think this looks good. Maybe you want to change the log message after a seek. I would assume that it always logs "unknown position after seek" now and the 'else' path is never taken. That feels like something went wrong.

**Comment by popcornmix** [comment]
@FernetMenta I noticed I wasn't setting `hasAudio`/`hasVideo` in the case where we had seen the stream, but `m_pkt.pkt.dts == AV_NOPTS_VALUE` so I've tweaked the logic to closer to the original code so that behaviour should no longer change.

Also edited the seek logging message.

**Comment by popcornmix** [comment]
jenkins failure looks spurious

jenkins build this please

**Comment by popcornmix** [comment]
We've got a thumbs up from @FernetMenta and a positive report this fixes reported issue.
Any other comments, or should we merge?

**Comment by enen92** [comment]
Go for it. Preferably in time for beta 3 for higher visibility/testing

**Comment by Portisch** [comment]
I tested this patch on Nexus as this is affected as well by the ffmpeg PTS issue. Most after resume playback from NAS after resume from suspend.
I did not have any Kodi freeze/deadlock anymore. Before it was about 2-3 times per week. But now like every second day without Kodi restart sometimes on channel change with PVR DVBviewer there is only GUI visible. Audio is there but no video. Stop/play the channel again fix it. But at least no Kodi deadlock anymore. So it's an improvement but there is still something missing.

**Comment by enen92** [comment]
I think this is causing regressions for me too, all my Pvr channels (hdhomerun) only have audio now:
https://paste.kodi.tv/yekubaxazu

Vs before:
https://paste.kodi.tv/izopugadel

**Comment by popcornmix** [comment]
The current code will set `hasVideo` only when a packet from video stream has been seen.
After your patch `hasVideo` will be set as long as there is a video stream, even if we have seen no packets.

If `hasVideo` is set we need to see a qualifying video packet (from video stream, with valid timestamp, and codec extradata seen) to start playback.

If `hasVideo` is unset then we are happy with seeing a qualifying audio packet (from audio stream, with valid timestamp) to start playback.

I think your patch makes sense to me. @FernetMenta are you happy this change is correct?

**Comment by FernetMenta** [comment]
> I think this is causing regressions for me too, all my Pvr channels (hdhomerun) only have audio now: https://paste.kodi.tv/yekubaxazu
> 
> Vs before: https://paste.kodi.tv/izopugadel

This is because hdhomerun does not send PAT/PMT when a stream is started and due to this change https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L2385 added video streams won't add the hasVideo flag anymore.

**Comment by FernetMenta** [comment]
@popcornmix the patch is not completely correct. Check my comment above. With mpegts a stream can appear at any time but the current code does not set hasVideo if m_seekStream was set already. The patch doesn't either.

**Comment by popcornmix** [comment]
I'll need to investigate that. I believe removing the early return when `m_seekStream` was set stopped seeking working (e.g. you seek forward by 10min and position is reported as, say, 3min).

**Comment by popcornmix** [comment]
I think the issue is that `IsTransportStreamReady()` returns false whenever a packet with no timestamp is read - there is no persistent state that says a stream is ready.

**Comment by enen92** [comment]
> @popcornmix the patch is not completely correct. Check my comment above. With mpegts a stream can appear at any time but the current code does not set hasVideo if m_seekStream was set already. The patch doesn't either.

what is the expected behaviour in case the `m_seekStream` is set and matches the current stream? Should it simply set hasVideo and wait for a valid packet to seen to mark the TS as ready?

Although I don't think it has any relation to my particular hdhomerun regression it'd be nice to know how you see it behaving on that case (not completely clear from your comment)

**Comment by FernetMenta** [comment]
@enen92 you need to deal with a case where no video stream is in the initial set of streams. So first you get an audio stream and set m_seekStream to this stream. When, at a later time, a video stream is detected, you need to set hasVideo. Means that an already set m_seekStream must not prevent hasVideo from being set. Further hasVideo must not be set until extradata is present.

As far as I remember hdhomerun has always passed an mpegts stream to kodi from the time you switched to that channel. That means that a program (PAT/PMT) may not be present at the beginning. That can trigger the case that hasVideo is not set.

**Comment by popcornmix** [comment]
@Portisch does https://github.com/xbmc/xbmc/pull/23629 help your issue?

**Comment by Portisch** [comment]
Thx, I am not able to test before end of next week.

**Pull request merged**